File: BotsSDK/korebot/src/main/java/kore/botssdk/audiocodes/webrtcclient/Fragments/ContactListFragment.java
package kore.botssdk.audiocodes.webrtcclient.Fragments;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ListView;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import kore.botssdk.R;
import kore.botssdk.audiocodes.webrtcclient.Adapters.ContactListAdapter;
import kore.botssdk.audiocodes.webrtcclient.General.ACManager;
import kore.botssdk.audiocodes.webrtcclient.General.Log;
import kore.botssdk.audiocodes.webrtcclient.Structure.ContactListObject;
import kore.botssdk.audiocodes.webrtcclient.db.NativeDBManager;
import kore.botssdk.audiocodes.webrtcclient.db.NativeDBObject;
import kore.botssdk.audiocodes.webrtcclient.db.NativeDBPhones;


public class ContactListFragment extends BaseFragment implements FragmentLifecycle {


    private final String TAG = "ContactListFragment";

    private ListView contactListView;
    private ContactListAdapter contactListAdapter;
    private int contactListMaxSize=3000;//0 for unlimited


    public ContactListFragment() {
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View rootView = inflater.inflate(R.layout.main_fragment_contact_list, container, false);

        initGui(rootView);


        return rootView;
        //return inflater.inflate(R.layout.main_fragment_dialer, container, false);
    }

    private void initGui(View rootView) {

        contactListView = (ListView) rootView.findViewById(R.id.contact_list_listview_contacts);
        //List<NativeContactObject> nativeContactObjectList = NativeContactUtils.getContactList(true);
        List<NativeDBObject> nativeDBObjectList = NativeDBManager.getContactList();

        Collections.sort(nativeDBObjectList);
        if (nativeDBObjectList != null)
        {
            Log.d(TAG, "original nativeDBObjectList: "+nativeDBObjectList.size());
            if (contactListMaxSize>0 && nativeDBObjectList.size()> contactListMaxSize)
            {
                nativeDBObjectList=nativeDBObjectList.subList(0,contactListMaxSize-1);
            }
        }


        List<ContactListObject> contactListObjectList = new ArrayList<ContactListObject>();
        // nativeContactObjectList.sort(Comparator.comparing(NativeContactObject::NativeContactObject.compare()));
        for (NativeDBObject nativeDBObject: nativeDBObjectList)
        {
            if(nativeDBObject!=null && nativeDBObject.getPhones()!=null && nativeDBObject.getPhones().size()>0)
            {
                for (NativeDBPhones nativeDBPhones:nativeDBObject.getPhones()) {
                    //.Log.d(TAG, "display name :"+nativeContactObject.getDisplayName());
                    contactListObjectList.add(new ContactListObject(nativeDBObject.getDisplayName(),nativeDBPhones.getPhoneNumber(),nativeDBObject.getPhotoURI(),nativeDBObject.getPhotoThumbnailURI()));
                }
            }
        }
        contactListAdapter = new ContactListAdapter(ContactListFragment.this.getContext(), contactListObjectList);

        contactListAdapter.setonCreateItemClickListener(clickListener);
        contactListAdapter.setCallButtonClickListener(callBtnClickListener);


        contactListView.setAdapter(contactListAdapter);



    }



    private ContactListAdapter.ListOnItemClickListener clickListener = new ContactListAdapter.ListOnItemClickListener() {
        @Override
        public void onItemClick(AdapterView<?> arg0, View v, int position, long id) {
            Log.d(TAG, "list view row selected");
            ContactListObject contactListObject = contactListAdapter.getList().get(position);
            Toast.makeText(ContactListFragment.this.getContext(), "audio call to: "+contactListObject.getName()+" number: "+contactListObject.getNumber(), Toast.LENGTH_SHORT).show();
            ACManager.getInstance().callNumber(contactListObject.getNumber());
        }
    };


    private ContactListAdapter.ListOnItemClickListener callBtnClickListener = new ContactListAdapter.ListOnItemClickListener() {
        @Override
        public void onItemClick(AdapterView<?> arg0, View v, int position, long id) {
            Log.d(TAG, "list view button selected");
            ContactListObject contactListObject = contactListAdapter.getList().get(position);
            Toast.makeText(ContactListFragment.this.getContext(), "Video call to: "+contactListObject.getName()+" number: "+contactListObject.getNumber(), Toast.LENGTH_SHORT).show();
            ACManager.getInstance().callNumber(contactListObject.getNumber(), true);

        }
    };

    @Override
    public void onPauseFragment() {

    }

    @Override
    public void onResumeFragment() {

    }

}

File: BotSdk/ui/src/main/java/com/kore/ui/audiocodes/webrtcclient/fragments/contactlist/ContactListFragment.kt
package com.kore.ui.audiocodes.webrtcclient.fragments.contactlist

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.LinearLayoutManager
import com.kore.ui.row.SimpleListAdapter
import com.kore.ui.row.SimpleListRow
import com.kore.common.utils.LogUtils
import com.kore.common.utils.ToastUtils
import com.kore.ui.R
import com.kore.ui.audiocodes.webrtcclient.db.NativeDBManager
import com.kore.ui.audiocodes.webrtcclient.fragments.BaseFragment
import com.kore.ui.audiocodes.webrtcclient.fragments.FragmentLifecycle
import com.kore.ui.audiocodes.webrtcclient.fragments.contactlist.row.ContactListItemRow
import com.kore.ui.audiocodes.webrtcclient.structure.ContactListObject
import com.kore.ui.databinding.MainFragmentContactListBinding

class ContactListFragment : BaseFragment(), FragmentLifecycle {
    companion object {
        private const val LOG_TAG = "ContactListFragment"
        private const val CONTACT_LIST_MAX_SIZE = 3000 //0 for unlimited
    }

    private var binding: MainFragmentContactListBinding? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = DataBindingUtil.inflate(layoutInflater, R.layout.main_fragment_contact_list, container, false)
        initGui()
        return binding?.root
    }

    private fun initGui() {
        binding?.apply {
            var nativeDBObjectList = NativeDBManager.getContactList(requireContext(), NativeDBManager.QueryType.ALL, null)
            LogUtils.d(LOG_TAG, "original nativeDBObjectList: ${nativeDBObjectList?.size}")
            if (nativeDBObjectList.isNullOrEmpty()) {
                ToastUtils.showToast(requireContext(), "No contacts found!", Toast.LENGTH_LONG)
                return@apply
            }
            val adapter = SimpleListAdapter(ContactListItemRowType.values().asList())
            contactList.layoutManager = LinearLayoutManager(root.context)
            contactList.addItemDecoration(SimpleListRow.VerticalSpaceItemDecoration(2))
            contactList.adapter = adapter
            if (nativeDBObjectList.size > CONTACT_LIST_MAX_SIZE) {
                nativeDBObjectList = nativeDBObjectList.subList(0, CONTACT_LIST_MAX_SIZE - 1)
            }
            val contactList: MutableList<ContactListObject> = ArrayList()
            for (dBObject in nativeDBObjectList) {
                dBObject?.phones?.map { dbPhone ->
                    if (dbPhone != null) {
                        contactList.add(
                            ContactListObject(dBObject.displayName, dbPhone.phoneNumber, dBObject.photoURI, dBObject.photoThumbnailURI)
                        )
                    }
                }
            }

            val rows = contactList.map {
                ContactListItemRow(it, this@ContactListFragment::onItemClick, this@ContactListFragment::onCallButtonClick)
            }
            adapter.submitList(rows)
        }
    }

    private fun onItemClick(contact: ContactListObject) {
        LogUtils.d(LOG_TAG, "list view row selected")
        Toast.makeText(requireContext(), "audio call to: ${contact.name} number: ${contact.number}", Toast.LENGTH_SHORT).show()
        acManager.callNumber(requireContext(), contact.number)
    }

    private fun onCallButtonClick(contact: ContactListObject) {
        LogUtils.d(LOG_TAG, "list view button selected")
        Toast.makeText(requireContext(), "Video call to: ${contact.name} number: ${contact.number}", Toast.LENGTH_SHORT).show()
        acManager.callNumber(requireContext(), contact.number, true)
    }

    override fun onPauseFragment() {}
    override fun onResumeFragment() {}
}